File: KIJUNG/PRO_WEEK_03/거리두기확인하기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <array>
#include <iostream>

#define FOR(i,a,b) for(int i=a;i<b;i++)

using namespace std;

int dRow[4] = {0,1,0,-1};
int dCol[4] = {1,0,-1,0};
vector<vector<char> > waitroom;
bool visited[5][5];

int DFS(int row, int col, int dist){
    if(waitroom[row][col] == 'P' && dist != 0){
        // cout << "row: " <<row << ", col: " << col << ", dist: " << dist << "\n";
        return 0;
    }

    FOR(i,0,4){
        if(dist >= 2) continue;
        if(row+dRow[i]<0 || row+dRow[i]>=5 || col+dCol[i]<0 || col+dCol[i]>=5) continue;
        if(visited[row+dRow[i]][col+dCol[i]]) continue;
        if(waitroom[row+dRow[i]][col+dCol[i]] == 'X') continue;
        
        visited[row+dRow[i]][col+dCol[i]] = true;
        if(DFS(row+dRow[i], col+dCol[i], dist+1) == 0) return 0;
        visited[row+dRow[i]][col+dCol[i]] = false;
    }

    return 1;

}

int check(vector<vector<char> > &waitroom){
    FOR(i,0,waitroom.size()){
        FOR(j,0,waitroom[i].size()){
            if(waitroom[i][j] == 'P'){
                visited[i][j] = true;
                if(DFS(i,j,0) == 0) return 0;
                visited[i][j] = false;
            }
        }
    }

    return 1;
}

vector<int> solution(vector<vector<string>> places) {
    vector<int> answer;
    
    for(vector<string> place: places) {
        FOR(i,0,5){
            vector<char> tmp;
            FOR(j,0,5){
                tmp.push_back(place[i][j]);
            }
            waitroom.push_back(tmp);
        }
        answer.push_back(check(waitroom));
        waitroom.clear();
    }
    return answer;
}

File: DONGUK/프로그래머스_WEEK 3/거리두기 확인하기.py
def search(place):
    # 상하좌우 1칸, 2칸, 대각선
    passed = [[0] * 5 for _ in range(5)]
    
    dx = [1, -1, 0, 0, 2, -2, 0, 0, 1, 1, -1, -1] 
    dy = [0, 0, 1, -1, 0, 0, 2, -2, 1, -1, 1, -1]
    for x in range(5):
        for y in range(5):
            if place[x][y] == 'P':
                for i in range(12):
                    nx = x + dx[i]
                    ny = y + dy[i]
                    
                    if 0 <= nx < 5 and 0 <= ny < 5:
                        if place[nx][ny] == 'P':
                            if passed[nx][ny] == 1:
                                continue
                            
                            if (abs(nx-x) + abs(ny-y)) == 1: # 붙어있는 경우
                                return False
                            
                            else:
                                if (x == nx) or (y == ny): # 같은 행 or 열
                                    if x == nx:
                                        if place[nx][(y+ny) // 2] == 'O':
                                            return False
                                    elif y == ny:
                                        if place[(x+nx)//2][ny] == 'O':
                                            return False
                                        
                                else: # 대각선
                                    if (nx-x) == 1 and (ny-y) == 1:
                                        if place[x+1][y] == 'X' and place[x][y+1] == 'X':
                                            continue
                                        else:
                                            return False
                                    
                                    elif (nx-x) == 1 and (ny-y) == -1:
                                        if place[x+1][y] == 'X' and place[x][y-1] == 'X':
                                            continue
                                        else:
                                            return False
                                        
                                    elif (nx-x) == -1 and (ny-y) == 1:
                                        if place[x-1][y] == 'X' and place[x][y+1] == 'X':
                                            continue
                                        else:
                                            return False
                                    
                                    elif (nx-x) == -1 and (ny-y) == -1:
                                        if place[x-1][y] == 'X' and place[x][y-1] == 'X':
                                            continue
                                        else:
                                            return False
            passed[x][y] = 1

    return True
                                    
                                            
    

def solution(places):
    answer = []
    
    for place in places:
        ret = search(place)
        print(ret)
    
    return answer